pytorch_lightning/trainer/training_loop.py

```diff
@@ -832,14 +832,14 @@ def build_train_args(self, batch, batch_idx, opt_idx, hiddens):
         # enable not needing to add opt_idx to training_step
         args = [batch, batch_idx]
 
-        if len(self.trainer.optimizers) > 1:
+        if len(self.trainer.optimizers) > 1 and self.automatic_optimization:
             if self.trainer.has_arg("training_step", "optimizer_idx"):
                 args.append(opt_idx)
             else:
                 num_opts = len(self.trainer.optimizers)
                 raise ValueError(
-                    f"Your LightningModule defines {num_opts} optimizers but "
-                    f'training_step is missing the "optimizer_idx" argument.'
+                    f"Your LightningModule defines {num_opts} optimizers but"
+                    f' training_step is missing the "optimizer_idx" argument.'
                 )
```

Comment:
with how dynamic args are appended here, what's the recommendation for customers who want to strongly type their lightning modules with either mypy or pyre? do we see this style as necessary in order to enable flexibility from the framework side?

Reply:
will be adding the support for old api with a deprecation warning.